### PR TITLE
Add support for custom HTTP post bodies via a template system

### LINKED ISSUE #1568: HTTP Post endpoints with custom HTTP Body
The HTTP Post Node should allow for the body to be customizable. This way Kapacitor will be able to inter-operate with many different systems since the body can be customized. 

Proposal:

* Add a `template` option to an httppost endpoint configuration that contains a Go template.
 Use a similar structure to alert messages to populate data for the template evaluation step.
* Add a `template_file` option which lets you define a template via an absolute file path instead of placing the template directly in the configuration.

@desa Thoughts?

### PR DESCRIPTION
**Pull request by nathanielc**
Fixes #1568 

- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated

### CODE REVIEW

**Comment by nathanielc** [comment]
@desa This is ready for review.

File: http_post.go [hunk 1]
```diff
@@ -132,7 +134,9 @@ func (n *HTTPPostNode) postRow(row *models.Row) {
 		return
 	}
 
-	req.Header.Set("Content-Type", "application/json")
+	if contentType != "" {
```
Comment (desa) [inline http_post.go]: Should `contentType` be settable from the config?

File: services/httppost/service.go [hunk 2]
```diff
@@ -266,8 +300,11 @@ func (h *handler) Handle(event alert.Event) {
 		return
 	}
 
+	if contentType != "" {
```
Comment (desa) [inline services/httppost/service.go]: Same as previous comment. Maybe contentType should be settable on an endpoint.

File: http_post.go (hunk 1 above)
Comment (nathanielc) [reply http_post.go]: It is already via the headers system.

Comment (desa) [reply http_post.go]: Ah good point.

**Comment by dishmael** [comment]
Is there documentation on how to use a template with custom JSON?

**Comment by nathanielc** [comment]
@dishmael Currently the docs around this feature exists in the configuration [here](https://github.com/influxdata/kapacitor/blob/master/etc/kapacitor/kapacitor.conf#L313) as the feature is controlled via the config.

**Comment by epsylonix** [comment]
Could you please clarify how this is supposed to work?

when doing a post from a .tick file with `alert().post('http://localhost:8084/kapacitor')` when there is a url defined in the kapacitor's `[[httppost]]` section, for example `url = "http://example.com"`, will a post be made to 'http://localhost:8084/kapacitor' or to 'http://example.com`?

I can't make this work unfortunately, when defining in the config file:
`[[httppost]]
alert-template = "{{.Message}}"`

I still get a default json format with all the default fields - details, id, etc. If I define both an inline template and a template file in the config file, I do get an error message so the config is loaded correctly, why isn't the custom template used? 

The same is true for a httpPost node, uncommenting `row-template = "{{.Name}} host={{index .Tags \"host\"}}{{range .Values}} {{index . \"time\"}} {{index . \"value\"}}{{end}}"` in the config file doesn't change the format data is posted in. 

Alert is written to a log file, so alert is generated correctly. There is no errors in the kapacitor log.

I'm new to the TICK stack, so I must be missing something, what else can I check? Thanks!

**Comment by epsylonix** [comment]
got it, so basically for a `httpPost()` the config data from the kapacitor's conf file is only used when an `endpoint` property is provided, different endpoints can have different configs, which config is used is determined by the endpoints name.

**Comment by dishmael** [comment]
@epsylonix I wasn't able to get this working properly. If you do, I would be interested in your approach/implementation.

**Comment by epsylonix** [comment]
@dishmael when you do a `post(url)` from `alertNode` or `httpRequest(url)` then the json with default fields is sent (fields are different for post and httpPost). To make use of this format feature you have to 

1. add a `[[httpPost]]` section in the .conf file. `[[` in toml means it's an array of objects, so there can be multiple `[[httpPost]]` sections. In that section you should define a url where data shold be posted, `alert-template` (used for `post()`), or/and `row-template` (for `httpPost()`), and the endpoint property, which is used as a key to reference this endpoint config fro a .tick file.

2. in the .tick file sue `alert().post().endpoint('foo')` or `httpPost().endpoint('foo')`, if you have `endpoint = "foo"` in the .conf file the config for that endpoint will be used, including a template.

By the way I didn't end up using this feature because it is not very flexible  - your template is a string with interpolation, so if you want json, you have to format it manually, and you can't pass random variables there - you have control over `.Message` and `.ID` or you can add tags in the alert data but it is not very convenient.

There is an `exec` property in `alertNode` and you can use any script with it and pass variables there as arguments, the alert data is written to STDIN. In that script it is easy to transform data and send it over http or any over channel in any format you like using you language of choice.

**Comment by dishmael** [comment]
Sadly, I tried all of that and could not get it to send an event.  I did not get anything in the log file that indicated an error associated with the settings so gave up.

**Comment by plalwa** [comment]
can we get a sample exec to pass any additional variables

**Comment by crashedserver** [comment]
@nathanielc A working example would be of great help

**Comment by caengcjd** [comment]
good job

**Comment by ednjv** [comment]
Seems like there's not much documentation around some hidden stuff here, also the default example doesn't work for me, it doesn't output anything for both `row-template` or `row-template-file`

I noticed that there's a custom golang `json` function inside the template, I can't see that being documented anywhere

**Comment by ashishkaransingh** [comment]
I used alert-template-file it works, also i can send hard coded tags (Hard_coded_Tag).
See sample post.html
{
"id": "{{.ID}}",
"message": "{{.Message}}:{{range .Data.Series}}{{.Tags}},{{range .Values}}{{.}}{{end}}{{end}}",
"Hard_coded_Tag": "TEST_007"
}

But i cannot use just {{.Tags}} or {{ index .Fields "value" }}?
Example:
{
"id": "{{.ID}}",
"message": "{{.Message}}:{{.Tags}}",
"Site_ID": "TEST_01"
}
Or
{
"id": "{{.ID}}",
"message": "{{.Tags}}",
"Site_ID": "TEST_01"
}

I get an error "httppost failed to execute alert template"

Any idea why

**Comment by ednjv** [comment]
@ashishkaransingh Alert templates don't have access to .Tags it's probably that, read here https://docs.influxdata.com/kapacitor/v1.5/event_handlers/post/#alert-templates

**Comment by ashishkaransingh** [comment]
@ednjv Testing with row-template
Has anyone got a way out of this?

**Comment by ednjv** [comment]
@ashishkaransingh I have take a look at this https://github.com/influxdata/kapacitor/issues/2149#issuecomment-479462608

**Comment by ashishkaransingh** [comment]
@ednjv  It was promising :) 
I tried this in kapacitor.conf 
[[httppost]]
  endpoint = "testendpoint"
  url = "https://..............................................."
  headers = {Content-Type = "application/json; charset=UTF-8"}
  #alert-template = ""
  #alert-template-file = "/temp/template/post.html"
  row-template = "{name_1: {{.Name}}}"

does not output {{.Name}}?

**Comment by ednjv** [comment]
@ashishkaransingh you need to use double quotes and escape them like this:

row-template = "{\"name_1\": \"{{.Name}}\"}"

**Comment by ashishkaransingh** [comment]
It did not work.
Is it like it only worked for you using kapacitor get/post api?

I used it in kapacitor.conf

[[httppost]]
  endpoint = "mytestendpoint"
  url = "https://testserver.net.................."
  headers = {Content-Type = "application/json; charset=UTF-8"}
  alert-template = ""
  ##alert-template-file = "/temp/template/post.html"
  row-template = "{\"name_1\": \"{{.Name}}\"}"

and Using .tick file where i have defined the endpoint:

var trigger = data
    |alert()
        .post()
        .endpoint('mytestendpoint')
        // .details('Details Removed')
        .crit(lambda: "value" > crit)
        .id(idVar)
        .idTag(idTag)

**Comment by ashishkaransingh** [comment]
@ednjv Unfortunately it did not work either.

**Comment by ednjv** [comment]
@ashishkaransingh wait, you are mixing stuff, you are mixing alert() with row-templates, if you want to use alert you should use alert-template instead of row-template

But if you want to use row-template you can't use alerts

**Comment by ashishkaransingh** [comment]
Ok, i see your point.
So far i am only able to make alert-template-file work.

But anyway it seems it is much easier to use variables inside tickscripts.
I think this depicts the request perfectly https://github.com/influxdata/kapacitor/issues/894

**Comment by y-ard** [comment]
Hi Guys,

Maybe I'll ask strange question, but do you know howto create custom json fields, by extracting value from the '.Data'. It is available as a template, but i don't know howto extract tags' values from it.
Any ideas?

**Comment by reactcker** [comment]
I tried something like the following but cannot access .Tags.
Was able to access {{.Data}} but could not extract key values.
I am sure will not be able to access fields ex: {{index .Fields "value"}}'

Can someone please confirm if we can make it work using alert templates or will this be possible in future with flux etc?

```
{
"id": "{{.ID}}",
"message": "{{.Message}}",
"details": "{{.Details}}",
"time": "{{.Time}}",
"duration": "{{ .Duration}}",
"level": "{{.Level}}",
"recoverable": {{.Recoverable}},
"data": {{ .Data }},
"Instance": "{{.Tags}}"
}

```
I tried this as well: "Instance": "{{index .Tags "instance"}}" same error as the one you here.

received error:

> ts=2020-03-27T21:29:10.006+05:30 lvl=error msg="failed to execute alert template" service=httppost task=http_post_alert_template_test err="template: body:10:15: executing \"body\" at <.Tags>: can't evaluate field Tags in type alert.Data"

**Comment by ashishkaransingh** [comment]
I tried something like the following but cannot access .Tags.
Was able to access {{.Data}} but could not extract key values.
I am sure will not be able to access fields ex: {{index .Fields "value"}}'

Can someone please confirm if we can make it work using alert templates or will this be possible in future with flux etc?

```
{
"id": "{{.ID}}",
"message": "{{.Message}}",
"details": "{{.Details}}",
"time": "{{.Time}}",
"duration": "{{ .Duration}}",
"level": "{{.Level}}",
"recoverable": {{.Recoverable}},
"data": {{ .Data }},
"Instance": "{{.Tags}}"
}

```
I tried this as well: "Instance": "{{index .Tags "instance"}}" same error as the one you here.

received error:

> ts=2020-03-27T21:29:10.006+05:30 lvl=error msg="failed to execute alert template" service=httppost task=http_post_alert_template_test err="template: body:10:15: executing \"body\" at <.Tags>: can't evaluate field Tags in type alert.Data"

**Comment by desa** [comment]
@ashishkaransingh this is the data that is available for the message https://docs.influxdata.com/kapacitor/v1.5/working/alerts/#alert-event-data

**Comment by ashishkaransingh** [comment]
> @ashishkaransingh this is the data that is available for the message https://docs.influxdata.com/kapacitor/v1.5/working/alerts/#alert-event-data

Thank you @desa!

I was looking to access/extract just .Tags or Vars 
It was important for us as we will integrate it with servicenow itom.. any help or if you can point me to golang templating and to the code to achieve this then i can contribute to it...

**Comment by ashishkaransingh** [comment]
> @dishmael when you do a `post(url)` from `alertNode` or `httpRequest(url)` then the json with default fields is sent (fields are different for post and httpPost). To make use of this format feature you have to
> 
> 1. add a `[[httpPost]]` section in the .conf file. `[[` in toml means it's an array of objects, so there can be multiple `[[httpPost]]` sections. In that section you should define a url where data shold be posted, `alert-template` (used for `post()`), or/and `row-template` (for `httpPost()`), and the endpoint property, which is used as a key to reference this endpoint config fro a .tick file.
> 2. in the .tick file sue `alert().post().endpoint('foo')` or `httpPost().endpoint('foo')`, if you have `endpoint = "foo"` in the .conf file the config for that endpoint will be used, including a template.
> 
> By the way I didn't end up using this feature because it is not very flexible - your template is a string with interpolation, so if you want json, you have to format it manually, and you can't pass random variables there - you have control over `.Message` and `.ID` or you can add tags in the alert data but it is not very convenient.
> 
> There is an `exec` property in `alertNode` and you can use any script with it and pass variables there as arguments, the alert data is written to STDIN. In that script it is easy to transform data and send it over http or any over channel in any format you like using you language of choice.

Can someone please provide an example to transform data using exec?

Thanks

**Comment by ashishkaransingh** [comment]
Good Tidings!
For anyone who is interested in accessing .tags for alerting let me know i found a way...

**Comment by ashishkaransingh** [comment]
1)
Use |alert node & httpPost() node in your tickscript Batch/Stream it works with both:
Note you can also add custom tags.. by using default() node.

 |alert()
        .details('details')
        .crit(lambda: "cpuusage" > crit)
        .message(message)
        .id(idVar)
        .idTag(idTag)
        .levelTag('level')
        .messageField(messageField)
        .durationField(durationField)
    |default()
        .tag('folder_name', 'new_sd343f')
        .tag('dashboard_name', 'cpu_ds')
        .tag('metric_name', 'cpuusage')
    |httpPost()
        .endpoint('vnexttopulse')
        .captureResponse()
2)
Use this in your kapacitor.conf

[[httppost]]
  endpoint = "someendpoint"
  url = "http://localhost:8081"
  headers = {Content-Type = "application/json"}
  row-template = '{"records":[{"source": "from_my_app", "event_class": "Kapacitor", "node": "{{index .Tags "host"}}", "type": "{{.Name}}", "metric_name": "{{index .Tags "metric_name"}}", "resource": "{{index .Tags "instance"}}", "additional_info": "Host: {{index .Tags "host"}} Instance: {{index .Tags "instance"}}", "description": "Browse Deeplink: (https://testsite.com/{{index .Tags "folder_name"}}/{{index .Tags "dashboard_name"}}?refresh=1m&orgId=1&var-host={{index .Tags "host"}}) {{range .Values}} Alert_Time: {{index . "time"}} Last_Measure: {{index . "cpuusage"}}" {{end}}}]}'
  [httppost.basic-auth]
    username = "xxxxxxxxxx"
    password = "yyyyyyyyyyyyy"

**Comment by ibnfalcon** [comment]
@ashishkaransingh I am using:

```
|httpPost()
.endpoint('test')

```

And chronograf is telling me that endpoint does not exist even though I have already declared it in the `kapacitor.conf` file. Can you please provide some feedback on this.


Thank you.

**Comment by reactcker** [comment]
@ibnfalcon 
can you can provide your how you got your [[httppost]] set up in kapacitor.conf
And it would be better if you post this to [InfluxData Community](https://community.influxdata.com/) I can answer there as well...

**Comment by y-ard** [comment]
agreed it's very confusing when someone is using chronograf 'httppost' configuration.
because chronograf isn't aware of this configuration in capacitor's config. Probably it should be a feature request for chronograf.

**Comment by ibnfalcon** [comment]
> @ibnfalcon
> What was exact error?
> Did you save and restart kapacitor?
> 
> can you can provide how you got your [[httppost]] set up in kapacitor.conf
> And it would be better if you post this to [InfluxData Community](https://community.influxdata.com/) I can answer there as well...

Eventually I was able to figure it out that if you make any changes to configuration, it will only when work if you restart. So yes, restarting fixed the issue.

Thank you.

**Comment by reactcker** [comment]
Great, You are welcome!

**Comment by ibnfalcon** [comment]
@reactcker 

Do you know if this syntax is supported in `alert-template-file`:

{
"id": "{{.ID}}",
"message": "{{.Message}}",
"time": "{{.Time}}",
`"level": "{{ if eq .Level "OK" }}Up{{ else }}Down{{ end }}",`
"data": {{ .Data }}
}


I am currently getting `error calling eq: incompatible types for comparison` because of `if else` condition for `.Level`.

**Comment by reactcker** [comment]
@ibnfalcon 
Yes,  syntax is supported.
However error points to "level": "{{ if eq .Level "OK" }}Up{{ else }}Down{{ end }}"

try something like this:
{{ if eq (.Level | print) "Warning" }}"3"{{else if eq (.Level | print) "OK"}}"0"{{ end }},

I strongly recommend you to put your topics of such nature on:
https://community.influxdata.com/

**Pull request merged**